### PR TITLE
feat(elementexplorer): highlight elements

### LINKED ISSUE #3465: Could not find testability for element when use elementor or element explorer
Hello, 

I am trying to run elementor, in order to locate elements on my page. But i always got the same message :

 ERROR: TypeError: Cannot create property 'stack' on string 'Error while waiting for Protractor to sync with the page: "Could not find testability for element."'

image

image

Same error when trying to using Element explorer. 

image
## Bug report
- Node Version: `6.2.0`
- Protractor Version: `4.0.3`
- Browser(s): `Chrome`
- Operating System and Version `Windows 7`
- Your protractor configuration file
- A relevant example test
- Output from running the test
- Steps to reproduce the bug
- The URL you are running your tests against (if relevant)
## Feature Request
- Reasons for adopting new feature
- Is this a breaking change? (How will this affect existing functionality)

### PR DESCRIPTION
**Pull request by cnishina**
- Highlight elements by changing css
- Use the rootEl default 'body'. On second attempt, use a default rootEl ''.
  After the attempt is made, reset the rootEl.

closes #3465

### CODE REVIEW

**Comment by cnishina** [comment]
Per conversation... should pull out the string function and add a js file.

**Comment by manoj9788** [comment]
Good to see this feature, however this was a simple hack that I am using currently: http://www.assertselenium.com/protractor/highlight-elements-during-your-protractor-test-run/

**Comment by juliemr** [comment]
This feels like two separate features - can I be a pain and ask you to split it into two PRs?

File: lib/browser.ts [hunk 1]
```diff
@@ -994,17 +999,9 @@ export class ProtractorBrowser extends Webdriver {
...
-          .browser.findElements(locator)
-          .then((arr: webdriver.WebElement[]) => {
-            let found: string[] = [];
-            for (let i = 0; i < arr.length; ++i) {
-              arr[i].getText().then((text: string) => { found.push(text); });
-            }
-            return found;
-          });
-    };
+    (<any>global).list = ExplorerScripts.list;
```
Comment (juliemr) [inline lib/browser.ts]: with `declare var global` at the top you shouldn't need the casting here.

Comment (juliemr) [inline lib/browser.ts]: (and throughout)

File: lib/browser.ts [hunk 2]
```diff
@@ -1168,16 +1176,16 @@ export class ProtractorBrowser extends Webdriver {
...
+      logger.info('------- Element Explorer ---------------------------------');
+      logger.info('Starting WebDriver debugger in a child process. Element');
+      logger.info('Explorer is still beta, please report issues at ');
+      logger.info('github.com/angular/protractor');
       logger.info();
       logger.info('Type <tab> to see a list of locator strategies.');
-      logger.info(
-          'Use the `list` helper function to find elements by strategy:');
-      logger.info('  e.g., list(by.binding(\'\')) gets all bindings.');
+      logger.info('Use the `list` and `highlight` helper function to find');
```
Comment (juliemr) [inline lib/browser.ts]: typo: helper functionS

File: lib/debugger/clients/explorerScripts.ts [hunk 3]
```diff
@@ -0,0 +1,88 @@
+import {promise, WebElement} from 'selenium-webdriver';
+
+import {ProtractorBrowser} from '../../browser';
+import {Locator} from '../../locators';
+let absoluteXPath = require('./explorerXpath').absoluteXPath;
+
```
Comment (juliemr) [inline lib/debugger/clients/explorerScripts.ts]: Use the `declare var global: any` trick here too.

File: lib/debugger/clients/explorerScripts.ts [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+  static highlight(locator: Locator) {
+    let xPaths: string[] = [];
+    let cssBorders: string[] = [];
+
+    return (<any>global)
+        .browser.findElements(locator)
+        .then((arr: WebElement[]) => {
+
+          for (let i = 0; i < arr.length; ++i) {
+            ExplorerScripts.getAbsoluteXPath(arr[i]).then(
```
Comment (juliemr) [inline lib/debugger/clients/explorerScripts.ts]: I think we can get around the whole xpath business by passing in the web element. Webdriver's `ExecuteScript` command lets you pass in a WebElement as an argument, which is converted to the proper HTMLElement on the browser side.

File: lib/browser.ts [hunk 5]
```diff
@@ -1074,9 +1071,20 @@ export class ProtractorBrowser extends Webdriver {
       execute_: function(execFn_: Function) {
         this.execPromiseResult_ = this.execPromiseError_ = undefined;
 
-        this.execPromise_ = this.execPromise_.then(execFn_).then(
-            (result: Object) => { this.execPromiseResult_ = result; },
-            (err: Error) => { this.execPromiseError_ = err; });
+        (<any>global).tempEl = (<any>global).browser.rootEl;
```
Comment (juliemr) [inline lib/browser.ts]: It makes me nervous to add this here. I think this should be a change in the general logic for how we look for the root Element, not something specific to the debugger (which is not just used for elementExplorer, but also other REPLs if I'm remembering correctly.

**Comment by cnishina** [comment]
Closing this again. We'll get it in one day.

**Pull request closed without merging**